File: Subdomain Enumerator/subdomain_enum.py
#!/usr/bin/env python3
"""
Subdomain Enumerator - DNS brute-forcing + External APIs
"""

import dns.resolver
import requests
import argparse
import sys
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import List, Set

class SubdomainEnumerator:
    def __init__(self, domain: str, threads: int = 50, api_keys: dict = None):
        self.domain = domain
        self.threads = threads
        self.api_keys = api_keys or {}
        self.subdomains: Set[str] = set()
        
    def load_wordlist(self, wordlist_path: str = None) -> List[str]:
        """Load subdomain wordlist"""
        default_wordlist = [
            'www', 'mail', 'ftp', 'localhost', 'webmail', 'smtp', 'pop', 'ns1', 'webdisk',
            'ns2', 'cpanel', 'whm', 'autodiscover', 'autoconfig', 'm', 'imap', 'test',
            'ns', 'blog', 'pop3', 'dev', 'www2', 'admin', 'forum', 'news', 'vpn', 'ns3',
            'mail2', 'new', 'mysql', 'old', 'lists', 'support', 'mobile', 'mx', 'static',
            'docs', 'beta', 'shop', 'sql', 'secure', 'demo', 'cp', 'calendar', 'wiki',
            'web', 'media', 'email', 'images', 'img', 'download', 'dns', 'piwik', 'stats',
            'dashboard', 'portal', 'manage', 'start', 'info', 'apps', 'video', 'sip',
            'dns2', 'api', 'cdn', 'mssql', 'remote', 'server', 'ftp2', 'stage', 'vps'
        ]
        
        if wordlist_path:
            try:
                with open(wordlist_path, 'r') as f:
                    return [line.strip() for line in f if line.strip()]
            except FileNotFoundError:
                print(f"Wordlist not found, using default")
        return default_wordlist
    
    def brute_force_dns(self, subdomain: str) -> bool:
        """Brute force DNS query for a subdomain"""
        target = f"{subdomain}.{self.domain}"
        try:
            resolver = dns.resolver.Resolver()
            resolver.timeout = 2
            resolver.lifetime = 2
            answers = resolver.resolve(target, 'A')
            if answers:
                ip = answers[0].to_text()
                print(f"[DNS] Found: {target} -> {ip}")
                self.subdomains.add(target)
                return True
        except:
            pass
        return False
    
    def api_securitytrails(self) -> Set[str]:
        """Query SecurityTrails API"""
        if not self.api_keys.get('securitytrails'):
            return set()
        
        url = f"https://api.securitytrails.com/v1/domain/{self.domain}/subdomains"
        headers = {'APIKEY': self.api_keys['securitytrails']}
        
        try:
            response = requests.get(url, headers=headers, timeout=10)
            if response.status_code == 200:
                data = response.json()
                subdomains = [f"{sub}.{self.domain}" for sub in data.get('subdomains', [])]
                for sub in subdomains:
                    print(f"[SecurityTrails] Found: {sub}")
                return set(subdomains)
        except Exception as e:
            print(f"SecurityTrails API error: {e}")
        return set()
    
    def api_virustotal(self) -> Set[str]:
        """Query VirusTotal API"""
        if not self.api_keys.get('virustotal'):
            return set()
        
        url = f"https://www.virustotal.com/api/v3/domains/{self.domain}/subdomains"
        headers = {'x-apikey': self.api_keys['virustotal']}
        
        try:
            response = requests.get(url, headers=headers, timeout=10)
            if response.status_code == 200:
                data = response.json()
                subdomains = []
                for item in data.get('data', []):
                    sub = item.get('id', '')
                    if sub.endswith(self.domain):
                        subdomains.append(sub)
                        print(f"[VirusTotal] Found: {sub}")
                return set(subdomains)
        except Exception as e:
            print(f"VirusTotal API error: {e}")
        return set()
    
    def api_censys(self) -> Set[str]:
        """Query Censys API (simplified)"""
        if not self.api_keys.get('censys'):
            return set()
        
        # Censys v2 API requires more complex auth
        print("[Censys] API integration requires additional setup")
        return set()
    
    def enumerate(self, wordlist_path: str = None):
        """Main enumeration function"""
        print(f"[*] Starting subdomain enumeration for: {self.domain}")
        print(f"[*] Using {self.threads} threads")
        
        # Load wordlist
        wordlist = self.load_wordlist(wordlist_path)
        print(f"[*] Loaded {len(wordlist)} subdomain candidates")
        
        # API enumeration
        print("\n[*] Querying external APIs...")
        with ThreadPoolExecutor(max_workers=3) as executor:
            api_futures = [
                executor.submit(self.api_securitytrails),
                executor.submit(self.api_virustotal),
                executor.submit(self.api_censys)
            ]
            
            for future in as_completed(api_futures):
                self.subdomains.update(future.result())
        
        # DNS brute-force
        print(f"\n[*] DNS brute-force in progress...")
        found_count = 0
        with ThreadPoolExecutor(max_workers=self.threads) as executor:
            future_to_sub = {executor.submit(self.brute_force_dns, sub): sub for sub in wordlist}
            
            for future in as_completed(future_to_sub):
                if future.result():
                    found_count += 1
        
        # Results
        print(f"\n[+] Enumeration complete! Found {len(self.subdomains)} unique subdomains")
        if self.subdomains:
            print("\n[+] Subdomains found:")
            for sub in sorted(self.subdomains):
                print(f"  - {sub}")

def main():
    parser = argparse.ArgumentParser(description='Subdomain Enumerator')
    parser.add_argument('-d', '--domain', required=True, help='Target domain')
    parser.add_argument('-w', '--wordlist', help='Custom wordlist file')
    parser.add_argument('-t', '--threads', type=int, default=50, help='Number of threads')
    parser.add_argument('--st-api', help='SecurityTrails API key')
    parser.add_argument('--vt-api', help='VirusTotal API key')
    
    args = parser.parse_args()
    
    api_keys = {
        'securitytrails': args.st_api,
        'virustotal': args.vt_api
    }
    
    enumerator = SubdomainEnumerator(args.domain, args.threads, api_keys)
    enumerator.enumerate(args.wordlist)

if __name__ == "__main__":
    main()

File: Subdomain Enumerator/subdomain_enum.js
#!/usr/bin/env node

const dns = require('dns').promises;
const https = require('https');
const fs = require('fs');
const { program } = require('commander');

class SubdomainEnumerator {
    constructor(domain, threads = 50, apiKeys = {}) {
        this.domain = domain;
        this.threads = threads;
        this.apiKeys = apiKeys;
        this.subdomains = new Set();
    }

    loadWordlist(wordlistPath = null) {
        const defaultWordlist = [
            'www', 'mail', 'ftp', 'localhost', 'webmail', 'smtp', 'pop', 'ns1', 'webdisk',
            'ns2', 'cpanel', 'whm', 'autodiscover', 'autoconfig', 'm', 'imap', 'test',
            'ns', 'blog', 'pop3', 'dev', 'www2', 'admin', 'forum', 'news', 'vpn', 'ns3',
            'api', 'cdn', 'stage', 'secure', 'portal'
        ];

        if (wordlistPath) {
            try {
                const content = fs.readFileSync(wordlistPath, 'utf8');
                return content.split('\n')
                    .map(line => line.trim())
                    .filter(line => line && !line.startsWith('#'));
            } catch (err) {
                console.log(`Wordlist not found, using default`);
            }
        }
        return defaultWordlist;
    }

    async bruteForceDNS(subdomain) {
        const target = `${subdomain}.${this.domain}`;
        try {
            const resolver = new dns.Resolver();
            resolver.setServers(['8.8.8.8', '1.1.1.1']);
            const addresses = await resolver.resolve4(target);
            if (addresses && addresses.length > 0) {
                console.log(`[DNS] Found: ${target} -> ${addresses[0]}`);
                this.subdomains.add(target);
                return true;
            }
        } catch (err) {
            // Subdomain not found
        }
        return false;
    }

    async apiSecurityTrails() {
        if (!this.apiKeys.securitytrails) return;

        const options = {
            hostname: 'api.securitytrails.com',
            path: `/v1/domain/${this.domain}/subdomains`,
            headers: { 'APIKEY': this.apiKeys.securitytrails },
            timeout: 10000
        };

        return new Promise((resolve) => {
            https.get(options, (res) => {
                let data = '';
                res.on('data', chunk => data += chunk);
                res.on('end', () => {
                    if (res.statusCode === 200) {
                        try {
                            const json = JSON.parse(data);
                            json.subdomains?.forEach(sub => {
                                const full = `${sub}.${this.domain}`;
                                console.log(`[SecurityTrails] Found: ${full}`);
                                this.subdomains.add(full);
                            });
                        } catch (e) {}
                    }
                    resolve();
                });
            }).on('error', () => resolve());
        });
    }

    async apiVirusTotal() {
        if (!this.apiKeys.virustotal) return;

        const options = {
            hostname: 'www.virustotal.com',
            path: `/api/v3/domains/${this.domain}/subdomains`,
            headers: { 'x-apikey': this.apiKeys.virustotal },
            timeout: 10000
        };

        return new Promise((resolve) => {
            https.get(options, (res) => {
                let data = '';
                res.on('data', chunk => data += chunk);
                res.on('end', () => {
                    if (res.statusCode === 200) {
                        try {
                            const json = JSON.parse(data);
                            json.data?.forEach(item => {
                                if (item.id?.endsWith(this.domain)) {
                                    console.log(`[VirusTotal] Found: ${item.id}`);
                                    this.subdomains.add(item.id);
                                }
                            });
                        } catch (e) {}
                    }
                    resolve();
                });
            }).on('error', () => resolve());
        });
    }

    async enumerate(wordlistPath = null) {
        console.log(`[*] Starting subdomain enumeration for: ${this.domain}`);
        console.log(`[*] Using ${this.threads} threads`);

        const wordlist = this.loadWordlist(wordlistPath);
        console.log(`[*] Loaded ${wordlist.length} subdomain candidates`);

        // API enumeration
        console.log(`\n[*] Querying external APIs...`);
        await Promise.all([
            this.apiSecurityTrails(),
            this.apiVirusTotal()
        ]);

        // DNS brute-force with concurrency control
        console.log(`\n[*] DNS brute-force in progress...`);
        const queue = [...wordlist];
        let active = 0;
        let completed = 0;

        const worker = async () => {
            while (queue.length > 0) {
                const sub = queue.shift();
                await this.bruteForceDNS(sub);
                completed++;
                if (completed % 100 === 0) {
                    process.stdout.write(`\r[*] Progress: ${completed}/${wordlist.length}`);
                }
            }
        };

        const workers = Array(Math.min(this.threads, wordlist.length))
            .fill()
            .map(() => worker());
        
        await Promise.all(workers);
        console.log(); // New line after progress

        // Results
        console.log(`\n[+] Enumeration complete! Found ${this.subdomains.size} unique subdomains`);
        if (this.subdomains.size > 0) {
            console.log(`\n[+] Subdomains found:`);
            Array.from(this.subdomains).sort().forEach(sub => {
                console.log(`  - ${sub}`);
            });
        }
    }
}

// CLI interface
program
    .requiredOption('-d, --domain <domain>', 'Target domain')
    .option('-w, --wordlist <path>', 'Custom wordlist file')
    .option('-t, --threads <number>', 'Number of threads', parseInt, 50)
    .option('--st-api <key>', 'SecurityTrails API key')
    .option('--vt-api <key>', 'VirusTotal API key')
    .parse(process.argv);

const options = program.opts();

const apiKeys = {
    securitytrails: options.stApi,
    virustotal: options.vtApi
};

const enumerator = new SubdomainEnumerator(options.domain, options.threads, apiKeys);
enumerator.enumerate(options.wordlist).catch(console.error);
